patches/server/0580-Add-StructuresLocateEvent.patch

```diff
@@ -0,0 +1,221 @@
++++ b/src/main/java/net/minecraft/world/level/levelgen/feature/ConfiguredStructureFeature.java
+@@ -38,6 +38,7 @@ public class ConfiguredStructureFeature<FC extends FeatureConfiguration, F exten
+     public final HolderSet<Biome> biomes;
+     public final Map<MobCategory, StructureSpawnOverride> spawnOverrides;
+     public final boolean adaptNoise;
++    static { io.papermc.paper.world.structure.PaperConfiguredStructure.init(); } // Paper
+ 
+     public ConfiguredStructureFeature(F feature, FC config, HolderSet<Biome> biomes, boolean bl, Map<MobCategory, StructureSpawnOverride> map) {
+         this.feature = feature;
+diff --git a/src/test/java/io/papermc/paper/world/structure/ConfiguredStructureTest.java b/src/test/java/io/papermc/paper/world/structure/ConfiguredStructureTest.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..e44822771a04549da63360d0e64467ea685e18eb
+--- /dev/null
++++ b/src/test/java/io/papermc/paper/world/structure/ConfiguredStructureTest.java
+@@ -0,0 +1,89 @@
++package io.papermc.paper.world.structure;
++
++import net.minecraft.data.BuiltinRegistries;
++import net.minecraft.resources.ResourceKey;
++import net.minecraft.resources.ResourceLocation;
++import net.minecraft.server.Bootstrap;
++import net.minecraft.world.level.levelgen.feature.ConfiguredStructureFeature;
++import net.minecraft.world.level.levelgen.structure.BuiltinStructures;
++import org.bukkit.NamespacedKey;
++import org.bukkit.StructureType;
++import org.bukkit.craftbukkit.util.CraftNamespacedKey;
++import org.bukkit.support.AbstractTestingBase;
++import org.junit.AfterClass;
++import org.junit.BeforeClass;
++import org.junit.Test;
++
++import java.io.PrintStream;
++import java.lang.reflect.Field;
++import java.util.LinkedHashMap;
++import java.util.Map;
++import java.util.Objects;
++import java.util.StringJoiner;
++
++import static org.junit.Assert.assertEquals;
++import static org.junit.Assert.assertNotNull;
++import static org.junit.Assert.assertTrue;
++
++public class ConfiguredStructureTest extends AbstractTestingBase {
++
++    private static final Map<ResourceLocation, String> BUILT_IN_STRUCTURES = new LinkedHashMap<>();
++    private static final Map<StructureType, String> API_STRUCTURE_TYPES = new LinkedHashMap<>();
++    private static PrintStream out;
++
++    @BeforeClass
++    public static void collectStructures() throws ReflectiveOperationException {
++        out = System.out;
++        System.setOut(Bootstrap.STDOUT);
++        for (Field field : BuiltinStructures.class.getDeclaredFields()) {
++            if (field.getType().equals(ResourceKey.class)) {
++                BUILT_IN_STRUCTURES.put(((ResourceKey<ConfiguredStructureFeature<?, ?>>) field.get(null)).location(), field.getName());
++            }
++        }
++
++        for (Field field : StructureType.class.getDeclaredFields()) {
++            if (field.getType().equals(StructureType.class)) {
```

Comment:
does this need to be changed to account for References? also is the DEFAULTS map really necessary given we scan the fields and build the API_STRUCTURE_TYPES map for testing? another thought is would it be possible/reasonable to make the tests more generic for when we add more similar registries in the future?

Reply:
Oh actually, that's not needed anymore. And that's `StructureType` not `ConfiguredStructure`. I needed that when I was generating the code to create the instances of `COnfiguredStructure` but since that's not happening now, it can just be removed.